primitives/messages/src/lib.rs

```diff
@@ -590,23 +608,43 @@ impl Default for OutboundLaneData {
 	}
 }
 
-/// Calculate the number of messages that the relayers have delivered.
-pub fn calc_relayers_rewards<AccountId>(
+/// Calculate the total relayer reward that need to be paid at the source chain.
+///
+/// The `compute_reward` assumed to be a function that maps the `RewardAtSource` onto
+/// real reward that is paid to relayer. This function needs to compute and return
+/// the total reward for delivering given number of messages to the target (bridged)
+/// chain by relayer that has agreed to work for given reward.
+pub fn calc_relayers_rewards_at_source<AccountId, Reward>(
 	messages_relayers: VecDeque<UnrewardedRelayer<AccountId>>,
 	received_range: &RangeInclusive<MessageNonce>,
-) -> RelayersRewards<AccountId>
+	compute_reward: impl Fn(MessageNonce, RewardAtSource) -> Reward,
+) -> RelayersRewards<AccountId, Reward>
 where
 	AccountId: sp_std::cmp::Ord,
+	Reward: Copy + Saturating + Zero,
 {
 	// remember to reward relayers that have delivered messages
 	// this loop is bounded by `T::MAX_UNREWARDED_RELAYERS_IN_CONFIRMATION_TX` on the bridged chain
-	let mut relayers_rewards = RelayersRewards::new();
+	let mut relayers_rewards: RelayersRewards<_, Reward> = RelayersRewards::new();
 	for entry in messages_relayers {
+		// if relayer does not expect any reward, do nothing
+		if entry.messages.reward_per_message == 0 {
+			continue
+		}
+
+		// if we have already paid reward for delivering those messages, do nothing
 		let nonce_begin = sp_std::cmp::max(entry.messages.begin, *received_range.start());
 		let nonce_end = sp_std::cmp::min(entry.messages.end, *received_range.end());
-		if nonce_end >= nonce_begin {
-			*relayers_rewards.entry(entry.relayer).or_default() += nonce_end - nonce_begin + 1;
+		let new_confirmations = nonce_begin..=nonce_end;
+		let new_confirmations_count = new_confirmations.saturating_len();
+		if new_confirmations_count == 0 {
+			continue
 		}
+
+		// compute and update reward in the rewards map
+		let new_reward = compute_reward(new_confirmations_count, entry.messages.reward_per_message);
```

Comment:
> Looks like compute_reward() needs all the info in entry.messages. So could we send only one parameter to it (&DeliveredMessages) ?

But the `calc_relayers_rewards_at_source` has some logic too. I.e. we may receive duplicate confirmations and obviously we shall not pay rewarrd twice. So this `new_confirmations_count` already has some logic behind it. Better to leave it here

> Or do we need the compute_reward() method at all ? Do we want the reward computation to be customizable ? Could we just receive a max_reward_per_message parameter instead ?

Now we are just using balance type as a reward. But it could be changed e.g. to `MultiAssets` in the future, which would make it hard to do comparisons here.

Reply:
Ok, makes sense.